apps/sel4test-tests/src/tests/schedcontext.c

```diff
@@ -732,12 +732,19 @@ test_revoke_sched_context_on_call_chain(env_t env)
 DEFINE_TEST(SCHED_CONTEXT_0013, "Test revoking a scheduling context on a call chain",
             test_revoke_sched_context_on_call_chain, config_set(CONFIG_KERNEL_MCS) &&config_set(CONFIG_HAVE_TIMER))
 
+void sched_context_0014_helper_fn(_Atomic int *state)
+{
+    while (1) {
+        *state = *state + 1;
```

Comment:
Do you want this or `*state += 1`? The difference is whether the compiler generates either a load acquire and a separate store release or it generates an atomic increment I think (From some aarch64 compiler tests).

Reply:
It could really be `*state = 1` here. I wouldn't expect the compiler to generate different code though, as += 1 is just a shorthand.

If doing the increment, calling seL4_Yield() first would make the test more robust I think. So I think I'll either change it to = 1 or add the yield.

Reply:
> I wouldn't expect the compiler to generate different code though, as += 1 is just a shorthand.

I was looking at what _Atomic actually does to an int and an int* and on my aarch64 compiler (12.2.0), a = a + 1; generated different atomic intrinsic than a += 1. The first being a separate load acquire/store release, and the second was a looping load exclusive/store conditional

Reply:
Was that with the actual build? For debug build there is no optimisation at all enabled so the generated assembly is horrible. But for optimised builds I would expect the same code, though neither gcc nor clang do:

https://godbolt.org/z/qbzxqPYTz

There's some subtle difference between the two that eludes me. For some reason neither generate atomic instructions for `*num = *num + 1;`.

Reply:
Your godbolt link is what I was seeing too. += and ++ with _Atomic are treated as a __atomic_fetch_add() whereas a = a + 1 is treated as a sequence of `__atomic_load_n` -> add 1 -> `__atomic_store_n`. So there's ordering but the fetch and add and write back isn't atomic. 
```
        ldar    w8, [x0]
        add     w8, w8, #1
        stlr    w8, [x0]
```
ldar and stlr are "atomics" in that they're `__atomic_load_n(num, __ATOMIC_ACQUIRE)`  and `__atomic_store_n(num, val, __ATOMIC_RELEASE)`

Reply:
Ah, thanks, I missed that `ldar` and `stlr` have acquire/release semantics. As long as the store is release it shouldn't really matter for the test though.